service-workers/redirects/test.window.js-serviceworker.js

```diff
@@ -0,0 +1,11 @@
+let evilGlobalState = null;
+onfetch = e => {
+  if(e.request.url.endsWith("test.window.js-redirect.py")) {
+    e.respondWith(new Promise(async resolve => {
+      evilGlobalState = await fetch(e.request);
+      resolve(new Response("stage 1 completed\n"));
+    }));
```

Comment:
I think the above 4 lines could be:

```js
evilGlobalState = fetch(e.request);
e.respondWith(new Response("stage 1 completed\n"));
```

Reply:
Wouldn't that increase the chance evilGlobalState is null when the second fetch comes in?

Reply:
Yeah, but if that's your intent then use `waitUntil`

```js
event.waitUntil(new Promise(r => self.globalStateCollected = r));
```

And call `globalStateCollected()` once `evilGlobalState` is no longer needed.